Sources/SwiftLexicalLookup/ScopeSyntax.swift

```diff
@@ -0,0 +1,76 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Returns all names that `for` refers to at this syntax node.
+  ///
+  /// - Returns: An array of names referred to by `for` at this syntax node,
+  /// ordered by visibility. The order is from the innermost to the outermost
+  /// scope, and within each scope, names are ordered by their introduction
+  /// in the source code.
+  ///
+  /// Example usage:
+  /// ```swift
+  /// class C {
+  ///   var a = 42
+  ///
+  ///   func a(a: Int) {
+  ///     a // <--- lookup here
+  ///
+  ///     let a = 0
+  ///   }
+  ///
+  ///   func a() {
+  ///     // ...
+  ///   }
+  /// }
+  /// ```
+  /// When calling this function on the declaration reference `a` within its name,
+  /// the function returns the parameter first, then the identifier of the variable
+  /// declaration, followed by the first function name, and then the second function name,
+  /// in this exact order. The constant declaration within the function body is omitted
+  /// due to the ordering rules that prioritize visibility within the function body.
+  @_spi(Experimental) public func lookup(for name: String) -> [LookupName] {
```

Comment:
I introduced a new enum `LookupResult` that, as for now, has one case `fromScope` that holds `ScopeSyntax` and `[LookupName]` associated with that scope. With the `lookup` function result type now being `[LookupResult]`, I think this could be a nice structure to work with. We could easily add new cases for the result that could encapsulate more information and e.g. act as compiler flags or provide suggestions for shadowing. What do you think about it?